fedimint-derive/src/lib.rs

```diff
@@ -214,124 +220,115 @@ pub fn derive_encodable(input: TokenStream) -> TokenStream {
-                                }
-                            };
-                            Ok(decoded)
-                        }
-                    }
-                }
+#[allow(unused_variables, unreachable_code)]
+fn error(ident: &Ident, message: &str) -> TokenStream2 {
+    #[cfg(feature = "diagnostics")]
+    ident.span().unstable().error(message).emit();
+    #[cfg(not(feature = "diagnostics"))]
+    panic!("{message}");
+
+    TokenStream2::new()
+}
+
+fn derive_struct_decode(ident: &Ident, fields: &Fields) -> TokenStream2 {
+    let decode_block = derive_tuple_or_named_decode_block(quote! { #ident }, &fields);
+
+    quote! {
+        Ok(#decode_block)
+    }
+}
+
+fn derive_enum_decode(ident: &Ident, variants: &Punctuated<Variant, Comma>) -> TokenStream2 {
+    if variants.is_empty() {
+        return quote! {
+            Err(::fedimint_core::encoding::DecodeError::new_custom(anyhow::anyhow!("Enum without variants can't be instantiated")))
+        };
+    }
+
+    let match_arms = variants.iter().enumerate().map(|(variant_idx, variant)| {
+        let variant_ident = variant.ident.clone();
+        let decode_block =
+            derive_tuple_or_named_decode_block(quote! { #ident::#variant_ident }, &variant.fields);
+
+        quote! {
+            #variant_idx => {
+                #decode_block
             }
         }
-        #[allow(unreachable_code)]
-        syn::Data::Union(_) => {
-            #[cfg(feature = "diagnostics")]
-            ident
-                .span()
-                .unstable()
-                .error("Encodable can't be derived for unions")
-                .emit();
-            #[cfg(not(feature = "diagnostics"))]
-            panic!("Error: Encodable can't be derived for unions");
+    });
 
-            return TokenStream::new();
+    quote! {
+        let variant = <u64 as ::fedimint_core::encoding::Decodable>::consensus_decode(d, modules)? as usize;
+        let decoded = match variant {
+            #(#match_arms)*
+            _ => {
+                return Err(::fedimint_core::encoding::DecodeError::from_str("invalid enum variant"));
+            }
+        };
+        Ok(decoded)
+    }
+}
+
+fn is_tuple_struct(fields: &Fields) -> bool {
+    fields.iter().any(|field| field.ident.is_none())
+}
+
+// TODO: how not to use token stream for constructor, but still support both:
+//   * Enum::Variant
+//   * Struct
+// as idents
+fn derive_tuple_or_named_decode_block(constructor: TokenStream2, fields: &Fields) -> TokenStream2 {
+    if is_tuple_struct(fields) {
+        derive_tuple_decode_block(constructor, fields)
+    } else {
+        derive_named_decode_block(constructor, fields)
+    }
+}
+
+fn derive_tuple_decode_block(constructor: TokenStream2, fields: &Fields) -> TokenStream2 {
+    let field_names = fields
+        .iter()
+        .filter(|f| panic_if_ignored(f))
+        .enumerate()
+        .map(|(idx, _)| format_ident!("field_{}", idx))
```

Comment:
wrong order of enumerate and filter

same in other places.

Reply:
Good catch, I blindly copied this over!